File: src/main/java/com/letgo/book/application/changeTitle/ChangeTitleCommandHandler.java
package com.letgo.book.application.changeTitle;

import com.letgo.book.domain.*;
import com.letgo.shared.application.bus.command.CommandHandler;
import com.letgo.shared.application.event.DomainEventPublisher;

final public class ChangeTitleCommandHandler implements CommandHandler<ChangeTitleCommand> {
    private final BookRepository repository;
    private final BookFinder finder;
    private final DomainEventPublisher publisher;

    public ChangeTitleCommandHandler(BookRepository repository, BookFinder finder, DomainEventPublisher publisher) {
        this.repository = repository;
        this.finder = finder;
        this.publisher = publisher;
    }

    public void handle(ChangeTitleCommand command) {
        Book book = finder.find(BookId.create(command.id()));
        BookTitle newTitle = BookTitle.create(command.newTitle());
        if (!book.title().equals(newTitle)) {
            book.changeTitle(newTitle);
            repository.save(book);
            publisher.publish(book.retrieveEvents());
        }
    }
}

File: src/main/java/com/letgo/book/application/changeTitle/ChangeTitleCommandHandler.kt
package com.letgo.book.application.changeTitle

import com.letgo.book.domain.BookFinder
import com.letgo.book.domain.BookId
import com.letgo.book.domain.BookRepository
import com.letgo.book.domain.BookTitle
import com.letgo.shared.application.bus.command.CommandHandler
import com.letgo.shared.application.bus.event.DomainEventPublisher

class ChangeTitleCommandHandler(
    private val repository: BookRepository,
    private val finder: BookFinder,
    private val publisher: DomainEventPublisher,
) : CommandHandler<ChangeTitleCommand> {
    override fun handle(command: ChangeTitleCommand) {
        val book = finder.find(BookId(command.id))
        val newTitle = BookTitle(command.newTitle, command.occurredOn)
        if (book.canChangeTitle(newTitle)) {
            book.changeTitle(newTitle)
            repository.save(book)
            publisher.publish(book.retrieveEvents())
        }
    }
}
